### DISCUSSION TITLE
JSBSim Turbulence Simulation

### DISCUSSION THREAD

**Discussion by Dobid** [JSBSim-Team/jsbsim#953] [Q&A]
Hello!

I'm currently working on RL based control for fixed wing UAVs under disturbances and I would like to know if someone could explain to me how JSBSim generates turbulences.

Additional details coming from my investigations:
When looking at the code, I see that the scale lengths and amplitudes are implemented as shown in the cited reference (MIL-F-8785C)  and in the [Continuous gusts wikipedia](https://en.wikipedia.org/wiki/Continuous_gusts#Altitude_dependence) page.
https://github.com/JSBSim-Team/jsbsim/blob/e4aa2a9c873edaa8723955835e48bbcbaf195f98/src/models/atmosphere/FGWinds.cpp#L258-L280

The next lines are dedicated to declaring variables following the Yeager report's conventions:
https://github.com/JSBSim-Team/jsbsim/blob/e4aa2a9c873edaa8723955835e48bbcbaf195f98/src/models/atmosphere/FGWinds.cpp#L292-L307

Then follows the implementation of the Tustin model. 
**Q1**: Since we're already in a `switch turbType; case ttTustin` why do we need a `if(turbType == ttTustin)` condition ?
https://github.com/JSBSim-Team/jsbsim/blob/e4aa2a9c873edaa8723955835e48bbcbaf195f98/src/models/atmosphere/FGWinds.cpp#L311-L344

**Q2**: The next condition: `if(turbType == ttMilspec` is never being run no ? since it's in the `case ttTustin` ?
https://github.com/JSBSim-Team/jsbsim/blob/e4aa2a9c873edaa8723955835e48bbcbaf195f98/src/models/atmosphere/FGWinds.cpp#L346-L355
And by the way I find it odd that the `case ttMilspec` is empty:
https://github.com/JSBSim-Team/jsbsim/blob/e4aa2a9c873edaa8723955835e48bbcbaf195f98/src/models/atmosphere/FGWinds.cpp#L247-L250

**Q3** : Where is the link to the Dryden spectrum model ? I couldn't find any part in the FGWinds.cpp where the Dryden power spectral densities are implemented as presented in the [Dryden Wind Turbulence Model wikipedia page](https://en.wikipedia.org/wiki/Dryden_Wind_Turbulence_Model#Power_Spectral_Densities) or the [Matworks Simulink Documentation](https://fr.mathworks.com/help/aeroblks/drydenwindturbulencemodelcontinuous.html)

So the question sythesizing all the previous ones could be: **what is the difference between Tustin and Milspec turbulences and their relationship to the Dryden spectrum model ?**

Sorry for the very long post, and thanks in advance !

**Comment by seanmcleod** [comment]
Just a quick initial answer for some of your questions.

> Q1: Since we're already in a switch turbType; case ttTustin why do we need a if(turbType == ttTustin) condition ?
> Q2: The next condition: if(turbType == ttMilspec is never being run no ? since it's in the case ttTustin ?
> And by the way I find it odd that the case ttMilspec is empty:

What you're missing is that the `ttMispec` case isn't empty, there is no `break` statement after it. So in effect the code below is common to both `ttMilspec` and `ttTustin` and hence the reason for those conditional checks since there are parts of the code that are different.

**Reply by Dobid** [reply]
Alright, I see haha. Dang it 🥲

**Comment by Dobid** [comment]
As an extra question, to give more detail about my "what's the difference between Milspec and Tustin turbulences" question; here I am playing with the Tustin turbulence model, and according to the plots the turbulence linear velocities reach abnormally high velocities ~300 m/s (at least to me) in comparison with Milspec turbulences with same settings (severe, windspeed_at_20ft = 75 fps, + 22 m/s north east constant wind).
Tustin turbulences inner loop control (roll, pitch, Va | ignore two first plots):
![jsbpid_inloop_tustin75-6_wind16n16e-22](https://github.com/JSBSim-Team/jsbsim/assets/36502022/9fbacd72-bc52-4842-9cc0-d14c96aff0df)

Milspec turbulences inner loop control (roll, pitch, Va | ignore two first plots):
![jsbpid_inloop_milspec75-6_wind16n16e-22](https://github.com/JSBSim-Team/jsbsim/assets/36502022/acef8890-8911-4f5c-a293-7b91fa69a18b)

**Comment by Dobid** [comment]
Friendly Up !

**Reply by agodemar** [reply]
@Dobid Hello, can you give us more context on your work with JSBSim?

**Reply by Dobid** [reply]
Hello,
I'm working on RL based / PID controller comparison for flight control under disturbances, hence me trying to understand precisely how turbulences are generated and the difference between the turbulence models that JSBSim offers. 

1st question :
I'm trying to understand the taxonomy of turbulence generation and more particularly how the Milspec and Tustin turbulence models are related to the Dryden spectrum model since I don't see the code for Dryden spectrum power spectral densities implemented in the JSBSim code as in [Dryden Wind Turbulence Model wikipedia page](https://en.wikipedia.org/wiki/Dryden_Wind_Turbulence_Model#Power_Spectral_Densities) or the [Matworks Simulink Documentation](https://fr.mathworks.com/help/aeroblks/drydenwindturbulencemodelcontinuous.html). The only common code I found are the scale lengths computation but it seems to come from the [Continuous gusts](https://en.wikipedia.org/wiki/Continuous_gusts#Altitude_dependence) formalism. 
So I don't really get why there is (Dryden spectrum) next to the Milspec and Tustin models in the FGWinds in the documentation. Is the continuous gust model the same as Dryden spectrum model ? The namings kinda confuse me.
![image](https://github.com/JSBSim-Team/jsbsim/assets/36502022/aca94cb4-2bb0-4e9f-91f5-ac5ece3d30d0)

2nd question :
These plots describe the telemetry of the flight of a fixed wing UAV flying under severe Tustin turbulences while being PID controlled for its pitch, roll and airspeed. Looking at the wind speeds (bottom right hand corner) plot, it seems that the wind reaches abnormally high speeds (up to 300m/s for the down wind) which doesn't seem normal to me, especially when compared to the Milspec turbulence model (with the same settings) giving much more coherent orders of magnitude for turbulent wind speeds.
![image](https://github.com/JSBSim-Team/jsbsim/assets/36502022/8bca1939-79fe-4db5-bd75-d1e5968d3b84)

That's it, feel free to ask more details if needed, thanks for your reply !

**Comment by seanmcleod** [comment]
> I'm trying to understand the taxonomy of turbulence generation...

I don't have a definite answer for this off the top of my head. But did notice this bit of documentation in `FGWinds.h` which may partially help, or be a starting point.

https://github.com/JSBSim-Team/jsbsim/blob/c7cd8ef03508b7a8d58a717aec4255410897a2e4/src/models/atmosphere/FGWinds.h#L56-L160

**Reply by Dobid** [reply]
Ok, that's the documentation I took a screenshot of, in my last reply. I may not keep on searching about it since it took me a lot of time already and I'm not sure if diving for so long in the details of old military specification reports is going to be worth the time for my work.

As for my second question about the Tustin turbulence simulation. Can you reproduce similar results for the wind speeds under the severe turbulence setting and confirm that those turbulences are giving abnormally high wind speeds ?

**Comment by seanmcleod** [comment]
> I'm not sure if diving for so long in the details of old military specification reports is going to be worth the time for my work.

Short of one of the original coders happening to see your discussion here and answering it I'm not sure you have much of a choice, unless someone else decides to read through the specification documents to come up with answers.

> As for my second question about the Tustin turbulence simulation. Can you reproduce similar results for the wind speeds....

Haven't tried it yet.

**Reply by Dobid** [reply]
Thanks for your replies!
As an extra information for the reproductibilty of my 2nd question, I launched JSBSim by setting the `simulation/randomseed` property to 1.
